### PR TITLE
streaming: Add stream_plan_ranges_percentage

### LINKED ISSUE #14191: RFC: Streaming may be inefficient and slow due to a static division of the streaming plan into 10 chunks
*Installation details*
HEAD: 4592bbe18207e9b25d485ba37103de9a35574a1d

*Description*
Streaming work of every streaming plan is processed in 10 chunks on a node level. Meaning that shards will not move to the next chunk work till all shards on the node finished the work in the context of a previous chunk:

https://github.com/scylladb/scylladb/blob/master/dht/range_streamer.cc#L267

And the thing is that it will always be that some shards are going to finish faster than others and then wait for slower ones to catch up.

And the more the difference in the amount of data to stream (e.g. if there are large partitions or something else that is harder to process like large cells) the more time those fast shard would sit idle do nothing.

Here is screenshot that demonstrates the phenomena: this is a per-shard view on a node that is being decommissioned where I emphasized two distinctive time frames:
* Fast: where all shards are streaming
* Slow: where only two are streaming because others have finished their streaming

image

And the funny thing is that it may be other shards that would be "slow" in the next chunk.
Hence if shards would stream not in chunks but "the whole thing" it would make the whole streaming finish much faster because shards would waste much less time in the "waiting" state.

There are multiple ways to make things faster in a simple easy change: e.g. make the number of chunks configurable and live-updatable. And possibly set the default to 1.

A more involving change would require us recall why this chunking logic is there to begin with. And this is in order to be more efficient in streaming errors handling: if there are any we are re-streaming the whole chunk. So, having to re-stream only 1/10 is obviously much better than the whole thing.

So, instead of resolving the errors issue by chopping the streaming plan we could be less lazy and actually store the current streaming progress (since we atomic streaming unit is a token range) in some `system` table, and do it for every shard independently. And since AFAIK the steaming plan progresses over token ranges monotonously this storing can be coalesced too if we see that updating the system table for every next range is too heavy.

This would not only resolve "chunking" issue and would allow each shard push all the way till the completion as fast as possible in the presence of possible errors and node restarts, but will also open a huge range of opportunities for additional APIs, e.g.:
* Pausing/resuming streaming.
* Canceling streaming.
* Progress tracking.

### PR DESCRIPTION
**Pull request by asias**
This option allows user to change the number of ranges to stream in batch per stream plan.

Currently, each stream plan streams 10% of the total ranges.

With more ranges per stream plan, it reduces the waiting time between two stream plans. For example,

stream_plan1: shard0 (t0), shard1 (t1)
stream_plan2: shard0 (t2), shard1 (t3)

We start stream_plan2 after all shards finish streaming in stream_plan1. If shard0 and shard1 in stream_plan1 finishes at different time. One of the shards will be idle.

If we stream more ranges in a single stream plan, the waiting time will be reduced.

Previously, we retry the stream plan if one of the stream plans is failed. That's one of the reasons we want more stream plans. With RBNO and 1f8b529e08bc61 (range_streamer: Disable restream logic), the restream factor is not important anymore.

Also, more ranges in a single stream plan will create bigger but fewer sstables on the receiver side.

The default value is the same as before: 10% percentage of total ragnes.

Fixes #14191

### CODE REVIEW

**Comment by mykaul** [comment]
@asias - who needs to review this?

**Comment by mykaul** [comment]
@vladzcloudius - does that satisfy https://github.com/scylladb/scylladb/issues/14191 ?

**Comment by mykaul** [comment]
@asias - any tests for this? (dtest or elsewhere ) ?

**Comment by vladzcloudius** [comment]
> @vladzcloudius - does that satisfy #14191 ?

I'm not sure. Will need @asias help on this.
It does provide a workaround - one can set this new parameter to 100. 
However if there will be errors? 

The description of https://github.com/scylladb/scylladb/commit/1f8b529e08bc613ca250314b51bbd41f728c1d7a has "with RBNO if there will be errors then it will be able to resume" but how will it know where to resume from? Does RBNO track the current progress? 

If it does - then great.
If it doesn't, then repair will still have to read-a-lot just to catch up to where it left off.

**Comment by asias** [comment]
> > @vladzcloudius - does that satisfy #14191 ?
> 
> I'm not sure. Will need @asias help on this. It does provide a workaround - one can set this new parameter to 100. However if there will be errors?

We do not create a stream plan to restream in case of error any more since 1f8b529e08bc613ca250314b51bbd41f728c1d7a.

In case of error, we need to restart the whole process. With / without this patch, it does not change things regarding restream. 


> 
> The description of [1f8b529](https://github.com/scylladb/scylladb/commit/1f8b529e08bc613ca250314b51bbd41f728c1d7a) has "with RBNO if there will be errors then it will be able to resume" but how will it know where to resume from? Does RBNO track the current progress?
> 
> If it does - then great. If it doesn't, then repair will still have to read-a-lot just to catch up to where it left off.

With RBNO, we do not track the progress. Repair will skip sync data to the local node automatically if it already has it. It is going to be fast to go over those already synced ranges because it only calculate a checksum and transfers no data on wire.

**Comment by asias** [comment]
> @asias - any tests for this? (dtest or elsewhere ) ?

We have lots of tests in dtest for the streaming.

**Comment by vladzcloudius** [comment]
>  It is going to be fast to go over those already synced ranges because it only calculate a checksum and transfers no data on wire.

Well, a more precise way to describe the current state would probably be "it's going to be **faster**" - not fast. If you have streamed 10TB of data so far it would require you to re-read all that 10TB on all RF replicas in order to calculate checksums and sync on them before you send the first "new" byte on wire. And this is just in a context of a single node. And if you need to do restart it a few times e.g. because you keep on hitting and fixing some corrupted sstables...

Bottom line, @mykaul, while this patch is definitely a step into a right direction and will indeed allow us to increase a streaming speed in cases where there are no errors we still need the second part of https://github.com/scylladb/scylladb/issues/14191 to make our streaming really great.

**Comment by mykaul** [comment]
> > @asias - any tests for this? (dtest or elsewhere ) ?
> 
> We have lots of tests in dtest for the streaming.

But none that change the number of ranges?

File: db/config.cc [hunk 1]
```diff
@@ -533,6 +533,8 @@ db::config::config(std::shared_ptr<db::extensions> exts)
         "Throttles all streaming file transfer between the data centers. This setting allows throttles streaming throughput betweens data centers in addition to throttling all network stream traffic as configured with stream_throughput_outbound_megabits_per_sec.")
     , stream_io_throughput_mb_per_sec(this, "stream_io_throughput_mb_per_sec", liveness::LiveUpdate, value_status::Used, 0,
         "Throttles streaming I/O to the specified total throughput (in MiBs/s) across the entire system. Streaming I/O includes the one performed by repair and both RBNO and legacy topology operations such as adding or removing a node. Setting the value to 0 disables stream throttling")
+    , stream_plan_ranges_batch_percentage(this, "stream_plan_ranges_batch_percentage", liveness::LiveUpdate, value_status::Used, 10,
+        "Specify the percentage of ranges to stream in a single stream plan")
```
Comment (bhalevy) [inline db/config.cc:537]: 1. nit: s/ragnes/ranges/ in commit message

2. `stream_plan_ranges_batch_percentage` seems unclear.  Given its description, how about `stream_plan_ranges_percentage`?

File: dht/range_streamer.cc [hunk 2]
```diff
@@ -292,6 +291,8 @@ future<> range_streamer::stream_async() {
                     for (auto it = range_vec.begin(); it < range_vec.end();) {
                         ranges_to_stream.push_back(*it);
                         ++it;
+                        auto percentage = _db.local().get_config().stream_plan_ranges_batch_percentage();
```
Comment (bhalevy) [inline dht/range_streamer.cc:294]: Will this get the most updated value in case it was live updated? (just to make sure)

Comment (asias) [reply dht/range_streamer.cc:294]: Yes, it is from the _db and it is live update.

File: db/config.cc (hunk 1 above)
Comment (asias) [reply db/config.cc:537]: OK. The word batch is not going to help much. I will change to the short one as you suggested.

**Comment by asias** [comment]
> > > @asias - any tests for this? (dtest or elsewhere ) ?
> > 
> > 
> > We have lots of tests in dtest for the streaming.
> 
> But none that change the number of ranges?

I will add a test to test this new option explicitly.

**Comment by asias** [comment]
Changes in d2c45c8f26dfe3a3c6ba490eeebdb4e51b737759:

- fixed typo in commit
- renamed to stream_plan_ranges_percentage

**Review by bhalevy — APPROVED** [review]

File: db/config.cc (hunk 1 above)
Comment (avikivity) [reply db/config.cc:537]: Should use a fraction (float 0..1), not percentage, gives better control.

Comment (asias) [reply db/config.cc:537]: It gives the granularity of 0.01 already. I am not sure why the a more finer granularity like 0.001 or 0.0001 is needed. In practice, user will probably just use the value of 100%  if they ever want to change the default value of stream_plan_ranges_batch_percentage.

Comment (avikivity) [reply db/config.cc:537]: It's more consistent with other fractions. See reduce_cache_sizes_at, wasm_memory_cache_fraction,  background_writer_scheduling_quota, dynamic_snitch_badness_threshold, background_writer_scheduling_quota.

Comment (asias) [reply db/config.cc:537]: OK. I am convinced by the consistent argument. Changed in v2. Thanks.

**Comment by asias** [comment]
Changes in v2:
- The ration value  is between 0 and 1 now.
